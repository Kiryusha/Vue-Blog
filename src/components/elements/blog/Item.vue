<template lang="pug">
  .item
    .heading
      router-link.image-wrapper(
        v-if="data.previewPicture"
        :to="`${data.code}/`"
      )
        .image(
          :style="`background-image:url(${data.previewPicture});`"
        )
      .title {{ data.title }}
      .info {{ date }}, {{ author }}
    .text {{ data.previewText }}
    .controls
      Button(
        v-if="isAuthor"
        :view="'delete'"
        :type="'a'"
        @click="$modal.show('delete', { code: data.code })"
      )
      Button(
        v-if="isAuthor"
        :view="'edit'"
        :type="'a'"
        @click="$router.push(`/publish/${data.code}/`)"
      )
      Button(
        :view="'more'"
        :type="'a'"
        @click="$router.push(`/blog/${data.code}/`)"
      ) Подробнее
</template>

<script>
import { mapState } from 'vuex';
import Button from 'Components/elements/general/Button';
import formatDate from '@/helpers/formatDate';

export default {
  components: {
    Button,
  },
  props: {
    data: {
      type: Object,
      default: () => ({}),
    },
  },
  computed: {
    ...mapState({
      isAdmin: state => state.user.isAdmin,
      userId: state => state.user.id,
    }),
    date() {
      return formatDate(this.data.date);
    },
    author() {
      return this.data.username ? this.data.username : 'без автора';
    },
    isAuthor() {
      return this.isAdmin || this.data.userId === this.userId;
    },
  },
};
</script>

<style lang="stylus" scoped>
  .item
    card()

    *
      overflow hidden
      text-overflow ellipsis
      word-wrap break-word

    .title
      font-family 'Roboto Slab', sans-serif
      font-size 30px
      max-width 630px

    .info
      font-family 'Roboto Slab', sans-serif
      font-size 16px
      color #CF
      padding-top 10px

    .text
      padding-top 10px
      font-size 16px
      line-height 1.7

    .controls
      padding-top 20px
      display flex
      justify-content flex-end

      a + span,
      a + a
        margin-left 10px

    .image-wrapper
      display block
      text-decoration none
      margin-bottom 35px

    .image
      overflow hidden
      width 100%
      height 300px
      background-size cover
      background-position center
      border 2px solid gray

    +phone()

      .image-wrapper
        margin-bottom 15px

      .title
        font-size 24px

      .image
        height 175px

</style>
