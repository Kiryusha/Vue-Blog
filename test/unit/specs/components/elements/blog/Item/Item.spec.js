// Item.vue

import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Item from 'Components/elements/blog/Item';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('Item.vue', () => {
  const config = {
    localVue,
    mocks: {
      $route: {
        params: {
          code: 'test',
        }
      },
      $modal: {
        show: () => {},
      },
      $router: {
        push: () => {},
      },
    }
  };

  let storeConfig = {
    state: {
      user: {
        userId: '1a',
        isAdmin: false,
      },
    },
  };
  let store = new Vuex.Store(storeConfig);

  describe('component shallow rendering', () => {
    it('should match a snapshot', async () => {
      const wrapper = shallowMount(Item, { ...config, store });

      await wrapper.setProps({
        data: {
          title: 'test',
          date: new Date('01/01/01'),
          username: 'author',
          previewText: 'test',
        }
      })

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('lifecycle-driven methods calls', () => {
    it('should set author to user name on created', async () => {
      const wrapper = shallowMount(Item, { ...config, store });

      await wrapper.setProps({
        data: {
          title: 'test',
          date: new Date('01/01/01'),
          username: 'author',
          previewText: 'test',
        }
      })

      expect(wrapper.vm.author).toBe('author');
    });
  });
});
