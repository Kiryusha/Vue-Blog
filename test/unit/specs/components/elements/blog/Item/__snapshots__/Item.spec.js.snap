// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item.vue component shallow rendering should match a snapshot 1`] = `
<div class="item">
  <div class="heading">
    <!---->
    <div class="title">test</div>
    <div class="info">01 Января 2001, author</div>
  </div>
  <div class="text">test</div>
  <div class="controls">
    <button-stub></button-stub>
    <button-stub></button-stub>
    <button-stub></button-stub>
  </div>
</div>
`;
